Source/SwiftLintBuiltInRules/Rules/Idiomatic/PreferTypeCheckingRule.swift

```diff
@@ -0,0 +1,107 @@
+            Example("""
+            if foo is Bar {
+                doSomeThing()
+            }
+            """),
+            Example("""
+            if let bar = foo as? Bar {
+                foo.run()
+            }
+            """),
+        ],
+        triggeringExamples: [
+            Example("bar ↓as? Foo != nil"),
+            Example("""
+            if foo ↓as? Bar != nil {
+                doSomeThing()
+            }
+            """),
+        ],
+        corrections: [
+            Example("bar ↓as? Foo != nil"): Example("bar is Foo"),
+            Example("""
+            if foo ↓as? Bar != nil {
+                doSomeThing()
+            }
+            """): Example("""
+            if foo is Bar {
+                doSomeThing()
+            }
+            """),
+        ]
+    )
+}
+
+private extension PreferTypeCheckingRule {
+    final class Visitor: ViolationsSyntaxVisitor<ConfigurationType> {
+        override func visitPost(_ node: UnresolvedAsExprSyntax) {
+            if node.asWithQuestionMarkExprIsBeingComparedToNotNil() {
+                violations.append(node.asKeyword.positionAfterSkippingLeadingTrivia)
+            }
+        }
+    }
+
+    final class Rewriter: ViolationsSyntaxRewriter<ConfigurationType> {
+        override func visit(_ node: ExprListSyntax) -> ExprListSyntax {
+            guard
+                node.unresolvedAsExprIsBeingComparedToNotNil,
+                let unresolvedAsExpr = node.dropFirst().first,
+                let indexUnresolvedAsExpr = node.index(of: unresolvedAsExpr),
+                let typeExpr = node.dropFirst(2).first
+            else {
+                return super.visit(node)
+            }
+            correctionPositions.append(unresolvedAsExpr.positionAfterSkippingLeadingTrivia)
+            let elements = node
+                .with(
+                    \.[indexUnresolvedAsExpr],
+                    "is \(typeExpr.trimmed)"
+                )
+                .dropLast(3)
+            let newNode = ExprListSyntax(elements)
+                .with(\.leadingTrivia, node.leadingTrivia)
+                .with(\.trailingTrivia, node.trailingTrivia)
+            return super.visit(newNode)
+        }
+    }
+}
+
+private extension ExprSyntaxProtocol {
+    func asWithQuestionMarkExprIsBeingComparedToNotNil() -> Bool {
+        guard let node = self.as(UnresolvedAsExprSyntax.self),
+              let parent = parent?.as(ExprListSyntax.self),
+              let last = parent.last, last.is(NilLiteralExprSyntax.self) else {
+            return false
+        }
+        return node.questionOrExclamationMark?.tokenKind == .postfixQuestionMark
+            && parent.dropLast().last?.as(BinaryOperatorExprSyntax.self)?.operator.tokenKind == .binaryOperator("!=")
+    }
+}
+
+private extension ExprListSyntax {
+    var unresolvedAsExprIsBeingComparedToNotNil: Bool {
+        guard
+            let node = dropFirst().first?.as(UnresolvedAsExprSyntax.self),
```

Comment:
I don't think this covers cases like `2*x as? X != nil` which are rare but possible. I recommend setting `foldExpressions: true` in `@SwiftSyntaxRule` so you can work on a folded expression tree. That should simplify catching such cases.

Also, rewriting the syntax tree should become more straightforward if you start with the parent node, which is `InfixOperatorExprSyntax` with a `!=` `BinaryOperatorExprSyntax`. You can then replace the entire node and use the same entry point for both the visitor and the rewriter.

Reply:
I added such a case as an example.

By using `InfixOperatorExprSyntax`, I was able to simplify the code. Thank you!